File: coding-patterns/16-binary-search/115.py
# Order-agnostic Binary Search (Easy)

class Solution:
  def search(self, arr, key):
    start, end = 0, len(arr) - 1
    ascending = arr[start] < arr[end]
    while start <= end:
      mid = start + (end - start) // 2

      if key == arr[mid]:
        return mid

      if ascending:  
        if key < arr[mid]:
          end = mid - 1  
        else:  
          start = mid + 1 
      else: 
        if key > arr[mid]:
          end = mid - 1  
        else:  
          start = mid + 1  

    return -1  

File: coding-patterns/16-binary-search/115.cpp
// Order-agnostic Binary Search (Easy)

using namespace std;

#include <iostream>
#include <vector>

class Solution {
public:
  int search(const vector<int>& arr, int key) {
    int start = 0;
    int end = arr.size()-1;
    bool ascending = arr[start] < arr[end];

    while (start <= end) {
      int mid = start + (end - start) / 2;

      if (key == arr[mid])
        return mid;

      if (ascending) {
        if (key < arr[mid])
          end = mid - 1;
        else
          start = mid + 1;
      } else {
        if (key > arr[mid])
          end = mid - 1;
        else
          start = mid + 1;
      }
    }
    return -1; 
  }
};
